lib/rubygems/printable_uri.rb

```diff
@@ -0,0 +1,60 @@
+# frozen_string_literal: true
+
+require_relative 'uri_parser'
+
+class Gem::PrintableUri
+  def self.parse_uri(uri)
+    new(uri).parse_uri
+  end
+
+  def initialize(original_uri)
+    @original_uri = original_uri
+  end
+
+  def parse_uri
+    @original_uri = Gem::UriParser.parse_uri(@original_uri)
+    @uri = @original_uri.clone
+    redact_credential if valid_uri?
+
+    self
```

Comment:
Maybe have to return `self` here is a sign that we should do the redacting on `initialize`, remove `Gem::PrintableUri.parse_uri` and replace `Gem::PrintableUri.parse_uri` instances with `Gem::PrintableUri.new`. At this point I'm probably just nitpicking...

Reply:
This could be only me being too cautious but I like to keep the `initialize` to be responsible only for initializing the object, meaning no external function call, etc. It's refactored to remove self by [`52ed495` (#4868)](https://github.com/rubygems/rubygems/pull/4868/commits/52ed49563a070cc052b412759c2740767c75ba63). Please take a look, if still what you described reads better to you, let me know.

Reply:
I'm fine with keeping `initialize` to the minimum, my concern was that the instance returned by `Gem::PrintableUri` is not really "usable" unless you redact credentials first, that's why it felt to me like `initialize`'s job. Returning `self` from `Gem::PrintableUri.parse_uri` didn't bother me per-se, but https://github.com/rubygems/rubygems/pull/4868/commits/52ed49563a070cc052b412759c2740767c75ba63 doesn't bother me either :)

I wonder if it would make sense to actually merge this credential redaction functionality and the parsing stuff off `Gem::UriParser` into a `Gem::Uri` class. In that case, you could instantiate a `Gem::Uri` which would be a wrapper around `URI`, which could be useful for some situations, but also redact credentials from it when needed.

Anyways, I think we can try these improvements separately?

Reply:
I ended up merging the classes at https://github.com/rubygems/rubygems/pull/4868/commits/9581c2740a37ef12ac0351cb9f166e3d6aa448c1. I think it's simpler that way, let me know what you think @simi and @intuxicated.